nosnoc/model.py

```diff
@@ -0,0 +1,360 @@
+        if not self.p_time_var_val.shape == (opts.N_stages, n_p_time_var):
+            raise Exception(
+                "dimension of p_time_var_val and p_time_var mismatch.",
+                f"Expected shape: ({opts.N_stages}, {n_p_time_var}), "
+                f"got p_time_var_val.shape {self.p_time_var_val.shape}"
+                f"p_time_var {self.p_time_var}, p_time_var_val {self.p_time_var_val}")
+        # extend parameters for each stage
+        n_p = n_p_time_var + n_p_glob
+        self.p = ca.vertcat(self.p_time_var, self.p_global)
+        self.p_ctrl_stages = [ca.SX.sym(f'p_stage{i}', n_p) for i in range(opts.N_stages)]
+
+        self.p_val_ctrl_stages = np.zeros((opts.N_stages, n_p))
+        for i in range(opts.N_stages):
+            self.p_val_ctrl_stages[i, :n_p_time_var] = self.p_time_var_val[i, :]
+            self.p_val_ctrl_stages[i, n_p_time_var:] = self.p_global_val
+
+        self.dims = NosnocDims(n_x=n_x,
+                               n_u=n_u,
+                               n_sys=n_sys,
+                               n_c_sys=n_c_sys,
+                               n_f_sys=n_f_sys,
+                               n_p_time_var=n_p_time_var,
+                               n_p_glob=n_p_glob)
+
+        if opts.pss_mode == PssMode.STEWART:
+            if self.g_Stewart:
+                g_Stewart_list = self.g_Stewart
+            else:
+                g_Stewart_list = [-self.S[i] @ self.c[i] for i in range(n_sys)]
+
+            g_Stewart = casadi_vertcat_list(g_Stewart_list)
+
+        # create dummy finite element - only use first stage
+        if len(self.alpha) != 0:
+            theta, lam, mu, _, lambda_n, lambda_p = self.create_stage_vars(opts)
+            alpha = self.alpha
+        else:
+            theta, lam, mu, alpha, lambda_n, lambda_p = self.create_stage_vars(opts)
+
+        # setup upsilon
+        upsilon = []
+        if opts.pss_mode == PssMode.STEP and self.F is not None:
+            for ii in range(self.dims.n_sys):
+                upsilon_temp = []
+                S_temp = self.S[ii]
+                for j in range(len(S_temp)):
+                    upsilon_ij = 1
+                    for k in range(len(S_temp[0, :])):
+                        # create multiafine term
+                        if S_temp[j, k] != 0:
+                            upsilon_ij *= (0.5 * (1 - S_temp[j, k]) +
+                                           S_temp[j, k] * alpha[ii][k])
+                    upsilon_temp = ca.vertcat(upsilon_temp, upsilon_ij)
+                upsilon = ca.horzcat(upsilon, upsilon_temp)
+
+        # start empty
+        g_lift = ca.SX.zeros((0, 1))
+        g_switching = ca.SX.zeros((0, 1))
+        g_convex = ca.SX.zeros((0, 1))  # equation for the convex multiplers 1 = e' \theta
+        lambda00_expr = ca.SX.zeros(0, 0)
+        std_compl_res = ca.SX.zeros(1)  # residual of standard complementarity
+
+        z = ca.vertcat(casadi_vertcat_list(theta),
```

Comment:
This of course has to be in the same order as in the `FiniteElement`.
But we can add a note about it later.